core-common/src/main/java/org/glassfish/jersey/internal/inject/Providers.java

```diff
@@ -350,16 +352,23 @@ public T apply(final RankedProvider<T> input) {
 
     private static <T> List<ServiceHandle<T>> getServiceHandles(final ServiceLocator locator, final Class<T> contract,
                                                                 final Annotation... qualifiers) {
-
-        final List<ServiceHandle<T>> allServiceHandles = qualifiers == null
+        final List<ServiceHandle<T>> serviceHandles = qualifiers == null
                 ? locator.getAllServiceHandles(contract)
                 : locator.getAllServiceHandles(contract, qualifiers);
-
-        final ArrayList<ServiceHandle<T>> serviceHandles = new ArrayList<ServiceHandle<T>>();
-        for (final ServiceHandle handle : allServiceHandles) {
-            //noinspection unchecked
-            serviceHandles.add((ServiceHandle<T>) handle);
-        }
+        Collections.sort(serviceHandles, new Comparator<ServiceHandle<T>>() {
```

Comment:
I notice that the GetAllServiceHandles class calls this method twice, and adds both results to the same list, which I believe will result in a partially sorted list. Is this right?

Reply:
I'm not seeing where this is called twice?

Reply:
Line 337/338

Reply:
Yep I think you're right. The first call is with a qualifier, and the second is without, but the sorting is happening on the subsets rather than on the full thing - I believe it does the same on the other PR that has been merged

Reply:
Fair observation, I hadn't noticed on the other PR

Reply:
Updated